torch/csrc/autograd/engine.h

```diff
@@ -68,6 +72,28 @@ struct TORCH_API Engine {
   std::vector<std::function<void()>> final_callbacks_;
   // To protect reads and writes to final_callbacks_
   std::mutex post_callbacks_lock_;
+
+  struct ThreadPoolShared {
+    // Data structres used by the threads for executing reentrant backwards
+    // tasks. See Note [Reentrant backwards]
+    // Number of avaliable threads for processing new GraphTasks.
+    std::atomic_uint num_workers_;
+    // The threads will wait on work_ to be notified of GraphTasks
+    std::condition_variable work_;
+    // To protect reads and writes to graphtask_queue_ and num_workers_
```

Comment:
Well `num_workers_` is atomic, so if you're going to read/write to it only under this mutex, then you should make it a regular variable.